libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/VoiceInstructionsPrefetcher.kt

```diff
@@ -0,0 +1,168 @@
+package com.mapbox.navigation.ui.voice.api
+
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.base.trip.model.RouteProgress
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.core.directions.session.RoutesExtra
+import com.mapbox.navigation.core.directions.session.RoutesObserver
+import com.mapbox.navigation.core.directions.session.RoutesUpdatedResult
+import com.mapbox.navigation.core.lifecycle.MapboxNavigationApp
+import com.mapbox.navigation.core.lifecycle.MapboxNavigationObserver
+import com.mapbox.navigation.core.trip.session.RouteProgressObserver
+import com.mapbox.navigation.utils.internal.Time
+import com.mapbox.navigation.utils.internal.ifNonNull
+
+/**
+ * Class that is responsible for listening to relevant updates and triggering
+ * voice instructions predownloading when needed.
+ * Register and unregister with [MapboxNavigationApp.registerObserver]
+ * and [MapboxNavigationApp.unregisterObserver]
+ * or invoke [onAttached] and [onDetached] manually if you are not using [MapboxNavigationApp].
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+class VoiceInstructionsPrefetcher internal constructor(
+    private val observableTime: Int,
+    private val timePercentageToTriggerAfter: Double,
+    private val speechApi: MapboxSpeechApi,
+    private val nextVoiceInstructionsProvider: NextVoiceInstructionsProvider,
+    private val timeProvider: Time,
+) : MapboxNavigationObserver {
+
+    /**
+     * Creates [VoiceInstructionsPrefetcher] with default
+     * observableTime and timePercentageToTriggerAfter.
+     * See [DEFAULT_OBSERVABLE_TIME_SECONDS] and [DEFAULT_TIME_PERCENTAGE_TO_TRIGGER_AFTER].
+     *
+     * @param speechApi [MapboxSpeechApi] instances that's used to generate instructions
+     */
+    constructor(speechApi: MapboxSpeechApi) : this(
+        DEFAULT_OBSERVABLE_TIME_SECONDS,
+        DEFAULT_TIME_PERCENTAGE_TO_TRIGGER_AFTER,
+        speechApi
+    )
+
+    /**
+     * Creates [VoiceInstructionsPrefetcher] with custom
+     * observableTime and timePercentageToTriggerAfter.
+     *
+     * @param observableTime voice instructions will be predownloaded for `observableTime` seconds
+     *  of route ahead
+     * @param timePercentageToTriggerAfter voice instructions predownloading will not be triggered
+     *  if `timePercentageToTriggerAfter` seconds did not pass since last predownloading session
+     * @param speechApi [MapboxSpeechApi] instances that's used to generate instructions
+     */
+    constructor(
+        observableTime: Int,
+        timePercentageToTriggerAfter: Double,
+        speechApi: MapboxSpeechApi,
```

Comment:
```suggestion
    internal constructor(
        observableTime: Int,
        timePercentageToTriggerAfter: Double,
        speechApi: MapboxSpeechApi,
```

Not a blocker because the object is experimental. But in order to preserve backwards compatibility, these should use the `Options` pattern.

Are the `observableTime` `timePercentageToTriggerAfter` parameters within the scope of what we want to support? It seems we should choose these parameters ourselves.

Reply:
It's supposed to be public. But sure, Options is a better approach here.

> It seems we should choose these values ourselves.

We can and we have reasonable defaults. I just added a way to customize it because it costs nothing. But I agree that maybe it's not worth introducing additional complex customization, especially for the first iteration.